README.md
# Simple market maker
Simple market maker bot

This bot places limit orders in [tauros](https://tauros.io) crypto exchange querying [Bitso](https://bitso.com) orderbook as reference price.

## Parameters

Working parameters:
* `is_active`: Designates wether the bot is active (`true` or `false`)
* `refresh_rate`: Time in minutes for price updating (e.g. every 3 minutes)


Trading related parameters:
* `market`: Market to place the order (BTC-MXN, LTC-MXN, BCH-MXN and ETH-MXN are by now allowed)
* `spread`: The minimum spread to take in consideration for order placement (e.g. 3%). Negative values are allowed.
* `order_value`: The maximum value that the order can have (e.g. $10,000.00 MXN)
* `side`: Order side (buy or sell)

This parameters can be confirgued locally in the file `robots.json` or in a remote firebase realtime database. View `settings.py` file.


## Business Logic

Rule 1:
Allways look for 1st position in order book

Rule 2:
Increase spread as much as posible trying to keep first position in order book

Rule 3:
Never place an order bellow configured spread

Rule 4:
Use the maximum available funds but never surpase configured order value.

Rule 5:
Ignore tauros orders located at first places if its value is bellow $200.00 MXN

Rule 6:
Ignore bitso orders localed at first places if its value is bellow $500.00 MXN


## Running the bots

### Set environment variables

  1. Copy to `env.example` into `.env`

    cp env.example .env

  2. Edit values in `.env` depending on your preferences

    nano .env

### Create and activate a python virtual envirtonment

You must first install or verify that your computer has `python 3.8`, `pip` and `virtualenv`

    virtualenv -p python3 env && source env/bin/activate
    pip install -r requirements.txt

### Run the bots
Do not forget to configure your bots using the `robots.json` file or in your firebase realtime database.

    python3 main.py


## Email notificacions
If some of your wallets runs out of funds, an email can be sent to notice you. You can follow this [tutorial](https://realpython.com/python-send-email/) for creating a dedicated gmail account.


## Docker
### Create the image
    docker image build --tag tauros-bot .

### Run the container
    docker container run --env-file ./.env tauros-bot


CHANGELOG.md
# November 2021

- [bugs-8](https://trello.com/c/GiU9WT4u/9-modulenotfounderror-no-module-named-simplejson): KeyError: 'data' | for ask in tauros_order_book['data']['asks']

# October 2020

- [bugs-7](https://trello.com/c/GiU9WT4u/9-modulenotfounderror-no-module-named-simplejson): ModuleNotFoundError: No module named 'simplejson'

- [notifications-7](https://trello.com/c/GiU9WT4u/7-send-notification-if-order-amount-below-minimum): Send notification if order amount below minimum

- [documentation-6](https://trello.com/c/lXHPstJz/6-add-envexample): Add env.example

- [documentation-5](https://trello.com/c/uEPkQcfo/5-add-changelogmd): Add CHANGELOG.md



settings.py
# settings.py
import os
from os.path import join, dirname
from dotenv import load_dotenv

dotenv_path = join(dirname(__file__), '.env')
load_dotenv(dotenv_path)

ENVIRONMENT = os.environ.get("ENVIRONMENT")
TAUR_API_KEY = os.environ.get("TAUR_API_KEY")
TAUR_API_SECRET = os.environ.get("TAUR_API_SECRET")

SENDER_EMAIL = os.environ.get("SENDER_EMAIL")
EMAIL_PASSWORD = os.environ.get("SENDER_EMAIL_PASSWORD")
RECEIVER_EMAIL = os.environ.get("RECEIVER_EMAIL")

NOTIFICATIONS_ENABLED = os.environ.get("NOTIFICATIONS_ENABLED") == '1'

REFRESH_ORDER_RATE = 60 * 3  # In seconds

NOT_FUNDS_AWAITING_TIME = 10 # In minutes

MIN_SPREAD = 0.015 # means 1.5 %

ORDER_PRICE_DELTA = 1 # Order price will be 1 MXN upper/lower than the best one

USE_FIREBASE = os.environ.get("USE_FIREBASE") == '1'

FIREBASE_PROJECT_ID=os.environ.get("FIREBASE_PROJECT_ID")


trading_bot/tauros_api.py
import requests
import json
import time
import hmac
import hashlib
import base64
import simplejson


class TaurosPrivate:

    def __init__(self, key, secret, prod=True):
        self.key = key
        self.secret = secret
        self.base_url = 'https://api.tauros.io' if prod else 'https://api.staging.tauros.io'

    def _get_signature(self, path, data, nonce, method='post'):
        request_data = json.dumps(data, separators=(',', ':'))
        message = str(nonce) + method.upper() + path + str(request_data)
        api_sha256 = hashlib.sha256(message.encode()).digest()
        api_hmac = hmac.new(base64.b64decode(self.secret), api_sha256, hashlib.sha512)
        api_signature = base64.b64encode(api_hmac.digest())
        signature = api_signature.decode()
        return signature

    def _request(self, path, data={}, query_params={}, method='post'):
        nonce = str(int(1000*time.time()))
        signature = self._get_signature(
            path=path,
            data=data,
            nonce=nonce,
            method=method,
        )
        headers = {
            'Authorization': 'Bearer {}'.format(self.key),
            'Taur-Signature': signature,
            'Taur-Nonce': nonce,
            'Content-Type': 'application/json',
        }
        try:
            return requests.request(
                method=method,
                url=self.base_url+path,
                data=json.dumps(data),
                params=query_params,
                headers=headers,
            ).json()
        except simplejson.errors.JSONDecodeError:
            return {
                'success': False,
                'msg': 'Could not connect to api.tauros.io'
            }

    def place_order(self, order):
        path = '/api/v1/trading/placeorder/'
        return self._request(path=path, data=order)

    def get_orders(self, market=None):
        path = '/api/v1/trading/myopenorders/'
        params = {}
        if market:
            params['market'] = market
        return self._request(path=path, query_params=params, method='get')

    def close_order(self, order_id):
        path = '/api/v1/trading/closeorder/'
        data = {
            'id': order_id,
        }
        return self._request(path=path, data=data)

    def get_wallet(self, coin):
        path = '/api/v1/data/getbalance/'
        data = {
            'coin': coin,
        }
        return self._request(path=path, query_params=data, method='get')


class TaurosPublic():

    def __init__(self,prod=True):
        self.base_url = 'https://api.tauros.io/api' if prod else 'https://api.staging.tauros.io/api'

    def _request(self, path, params={}):
        try:
            return requests.get(
                url=self.base_url+path,
                params=params,
            ).json()
        except simplejson.errors.JSONDecodeError:
            return {
                'success': False,
                'msg': 'Could not connect to api.tauros.io'
            }

    def get_order_book(self, market='BTC-MXN'):
        path = f'/v2/trading/{market}/orderbook/'
        return self._request(path=path)



trading_bot/notifications.py
import smtplib, ssl
from email.message import EmailMessage
import settings

port = 465  # For SSL

# Create a secure SSL context
context = ssl.create_default_context()

def send_email(message):
    with smtplib.SMTP_SSL("smtp.gmail.com", port, context=context) as server:
        server.login(settings.SENDER_EMAIL, settings.EMAIL_PASSWORD)
        server.sendmail(
            settings.SENDER_EMAIL, settings.RECEIVER_EMAIL, message
        )

def send_funds_status_email(left_coin_balance, right_coin_balance, market):
    if not settings.NOTIFICATIONS_ENABLED:
        return
    left_coin, right_coin = market.split('-')
    msg = EmailMessage()
    msg.set_content(
        f'Reporte del estado de sus balances en tauros.io en el mercado {market} \n'
        f'• {left_coin_balance} {left_coin} \n'
        f'• {right_coin_balance} {right_coin}'
    )
    msg['Subject'] = f'{left_coin_balance} {left_coin}, {right_coin_balance} {right_coin}'
    msg['From'] = 'Tauros trading bot'
    msg['To'] = 'Marco Montero'
    send_email(msg.as_string())

main.py
import bitso
from trading_bot.tauros_api import TaurosPrivate, TaurosPublic
from trading_bot import notifications
from decimal import Decimal
import requests
import logging
import settings
import time
import json

from threading import Thread

logging.basicConfig(filename='logs.log', level=logging.ERROR)

tauros_key = settings.TAUR_API_KEY
tauros_secret = settings.TAUR_API_SECRET
is_production = settings.ENVIRONMENT == 'prod'

if not tauros_key or not tauros_secret:
    raise ValueError('Tauros credentials not fund')

tauros = TaurosPrivate(key=tauros_key, secret=tauros_secret, prod=is_production)

tauros_public = TaurosPublic(prod=is_production)

bisto_api = bitso.Api()

def close_all_orders():
    '''
    This function queries all open orders in tauros and closes them.
    '''
    open_orders = tauros.get_orders()
    if not open_orders['success']:
        logging.error(f'Querying open orders fail. Error: {open_orders["msg"]}')
        return

    orders_ids = [order['order_id']  for order in open_orders['data']]
    logging.info(f'Open orders: {orders_ids}')
    orders_closed = 0
    for order_id in orders_ids:
        print("Closing order with id: ", order_id)
        close_order = tauros.close_order(order_id=order_id)
        if not close_order['success']:
            print(f'Close order with id {order_id} failed. Error: ', close_order['msg'])
            continue
        orders_closed += 1
    print(f'{orders_closed} limit orders closed!')

def get_buy_order_price(max_price, ref_price, spread=None):
    if spread is None:
        spread =  settings.MIN_SPREAD
    max_price = max_price * Decimal(1 - spread / 100)
    if ref_price > max_price:
        return max_price
    return ref_price + settings.ORDER_PRICE_DELTA

def get_sell_order_price(min_price, ref_price, spread=None):
    if spread is None:
        spread =  settings.MIN_SPREAD
    min_price = min_price * Decimal(1 + spread / 100)
    if ref_price < min_price:
        return min_price
    return ref_price - settings.ORDER_PRICE_DELTA

def send_not_enough_balance_notification(left_coin_balance=None, right_coin_balance=None):
    if right_coin_balance is None:
        right_coin_wallet = tauros.get_wallet('mxn')
        right_coin_balance = right_coin_wallet['data']['balances']['available']

    if left_coin_balance is None:
        left_coin_wallet = tauros.get_wallet('btc')
        left_coin_balance = left_coin_wallet['data']['balances']['available']

    notifications.send_funds_status_email(
        left_coin_balance=left_coin_balance,
        right_coin_balance=right_coin_balance,
        market='BTC-MXN',
    )

def get_bitso_bid(market='btc-mxn'):
    # Getting bitso order book
    bitso_order_book = bisto_api.order_book(market.replace('-', '_'))

    bitso_price = None
    for bid in bitso_order_book.bids:
        bid_value = bid.price * bid.amount
        if bid_value >= Decimal('500.00'):
            bitso_price = bid.price
            return bitso_price


def get_bitso_ask(market='btc-mxn'):
    # Getting bitso order book
    bitso_order_book = bisto_api.order_book(market.replace('-', '_'))

    bitso_price = None
    for ask in bitso_order_book.asks:
        ask_value = ask.price * ask.amount
        if ask_value >= Decimal('500.00'):
            bitso_price = ask.price
            return bitso_price


def get_tauros_bid(market='btc-mxn'):
    # Getting tauros order book
    tauros_order_book = tauros_public.get_order_book(market=market)
    tauros_price = None
    for bid in tauros_order_book['payload']['bids']:
        if Decimal(str(bid['value'])) > Decimal('200.00'):
            tauros_price = Decimal(str(bid['price']))
            return tauros_price


def get_tauros_ask(market='btc-mxn'):
    # Getting tauros order book
    tauros_order_book = tauros_public.get_order_book(market=market)
    tauros_price = None
    for ask in tauros_order_book['payload']['asks']:
        if Decimal(str(ask['value'])) > Decimal('200.00'):
            tauros_price = Decimal(str(ask['price']))
            return tauros_price


def get_order_value(max_balance, price, max_order_value=200_00.00, side='buy'):
    # Setting order value
    MAX_ORDER_VALUE = Decimal(str(max_order_value))

    if side == 'buy':
        if max_balance > MAX_ORDER_VALUE:
            return MAX_ORDER_VALUE
        else:
            return max_balance

    order_value = max_balance * price

    if order_value > MAX_ORDER_VALUE:
        return MAX_ORDER_VALUE

    return order_value

def sell_bot(config_id, remote=False):
    config = {}
    while True:
        if remote:
            response = requests.get(f'https://{settings.FIREBASE_PROJECT_ID}.firebaseio.com/{config_id}.json')
            config = response.json()
        else:
            with open('./robots.json') as robots:
                robots_list = json.load(robots)
                config = robots_list[config_id]
        market = config['market']
        spread = config['spread']
        time_to_sleep = config.get('refresh_rate') * 60 or settings.REFRESH_ORDER_RATE

        if not config.get('is_active'):
            print(f"{market} bot is not active. Sleeping {time_to_sleep} seconds")
            time.sleep(time_to_sleep)
            continue

        left_coin, right_coin = market.split('-')
        bitso_price = get_bitso_ask(market=market)
        tauros_price = get_tauros_ask(market=market)

        if not bitso_price or not tauros_price:
            logging.error('Bitso or Tauros query price failed')
            time.sleep(3)
            continue

        left_coin_wallet = tauros.get_wallet(left_coin)

        if not left_coin_wallet['success']:
            logging.error(f'Tauros {left_coin} wallet query failed')
            logging.error(left_coin_wallet['msg'])
            time.sleep(3)
            continue

        left_coin_balance = Decimal(
            left_coin_wallet['data']['balances']['available']
        )

        if left_coin_balance == 0:
            logging.error(f'{left_coin} wallet is empty. Imposible to place a buy order. Sending email . . .')
            send_not_enough_balance_notification(left_coin_balance=0)
            time.sleep(settings.NOT_FUNDS_AWAITING_TIME * 60)
            continue

        order_price = get_sell_order_price(
            min_price=bitso_price,
            ref_price=tauros_price,
            spread=spread,
        )

        order_value = get_order_value(
            max_balance=left_coin_balance,
            price=order_price,
            max_order_value=config.get('order_value') or 20_000.00,
            side='sell'
        )

        order = {
            "market": market,
            "amount": str(order_value),
            "is_amount_value": True,
            "side": "SELL",
            "type": "LIMIT",
            "price": str(order_price),
        }

        order_placed = tauros.place_order(order=order)

        if not order_placed['success']:
            print('=======================================================')
            error_msg = f"Could not place sell order in {market} market. Error: {order_placed['msg']}"
            logging.error(error_msg)
            print(error_msg)
            messages = (
                'The minimum order',
                'has not enough {}'.format(left_coin.upper()),
                "'amount' field must be greater",
            )
            try:
                for message in messages:
                    if message in order_placed['msg'][0]:
                        send_not_enough_balance_notification()
                        logging.error("Not enough funds email sent...")
                        time.sleep(settings.NOT_FUNDS_AWAITING_TIME * 60)
            except:
                pass
            continue

        order_id = order_placed['data']['id']

        time_to_sleep = config.get('refresh_rate') * 60 or settings.REFRESH_ORDER_RATE
        print('=======================================================')
        print("Market: ", market)
        print('Side: SELL')
        print('Tauros ask order price: ', tauros_price)
        print('Bitso ask order price: ', bitso_price)
        print("Sell order successfully placed: ")
        order_data = order_placed['data']
        real_spread = (bitso_price - Decimal(order_data['price'])) / bitso_price
        real_spread = abs(round(real_spread * 100, 2))
        print(
            f"DT: {order_data['created_at']} | Price: {order_data['price']} | Spread: {real_spread}% | Amount: {order_data['amount']} | Value: {order_data['value']}"
        )
        print(f"Sleeping {time_to_sleep} seconds")
        time.sleep(time_to_sleep)

        close_order = tauros.close_order(order_id)
        if not close_order['success']:
            print("Order close faild. Error", close_order['msg'])
            close_all_orders()


def buy_bot(config_id, remote=False):
    config = {}
    while True:
        if remote:
            response = requests.get(f'https://{settings.FIREBASE_PROJECT_ID}.firebaseio.com/{config_id}.json')
            config = response.json()
            if not config:
                print("==================================================")
                print("Imposible to place order. Could get confi from firebase. Error:", respose.text)
                return
        else:
            with open('./robots.json') as robots:
                robots_list = json.load(robots)
                config = robots_list[config_id]
        market = config['market']
        spread = config['spread']
        time_to_sleep = config.get('refresh_rate') * 60 or settings.REFRESH_ORDER_RATE

        if not config.get('is_active'):
            print(f"{market} bot is not active. Sleeping {time_to_sleep} seconds")
            time.sleep(time_to_sleep)
            continue

        left_coin, right_coin = market.split('-')
        bitso_price = get_bitso_bid(market=market)
        tauros_price = get_tauros_bid(market=market)

        if not bitso_price or not tauros_price:
            logging.error('Bitso or Tauros query price failed')
            time.sleep(3)
            continue

        right_coin_wallet = tauros.get_wallet(right_coin)

        if not right_coin_wallet['success']:
            logging.error(f'Tauros {right_coin} wallet query failed')
            logging.error(right_coin_wallet['msg'])
            time.sleep(3)
            continue

        right_coin_balance = Decimal(
            right_coin_wallet['data']['balances']['available']
        )

        if right_coin_balance == 0:
            print('=======================================================')
            logging.error(f'{right_coin} wallet is empty. Imposible to place a buy order. Sending email . . .')
            send_not_enough_balance_notification(right_coin_balance=0)
            time.sleep(settings.NOT_FUNDS_AWAITING_TIME * 60)
            continue

        order_price = get_buy_order_price(
            max_price=bitso_price,
            ref_price=tauros_price,
            spread=spread,
        )

        order_value = get_order_value(
            max_balance=right_coin_balance,
            price=order_price,
            max_order_value=config.get('order_value') or 20_000.00,
            side='buy'
        )

        order = {
            "market": market,
            "amount": str(order_value),
            "is_amount_value": True,
            "side": "BUY",
            "type": "LIMIT",
            "price": str(order_price),
        }

        order_placed = tauros.place_order(order=order)

        if not order_placed['success']:
            print('=======================================================')
            error_msg = f"Could not place buy order in {market} market. Error: {order_placed['msg']}"
            logging.error(error_msg)
            print(error_msg)
            messages = (
                'The minimum order',
                'has not enough {}'.format(right_coin.upper()),
                "'amount' field must be greater",
            )
            try:
                for message in messages:
                    if message in order_placed['msg'][0]:
                        send_not_enough_balance_notification()
                        logging.error("Not enough funds email sent...")
                        time.sleep(settings.NOT_FUNDS_AWAITING_TIME * 60)
            except:
                pass
            continue

        order_id = order_placed['data']['id']

        print('=======================================================')
        print("Market: ", market)
        print('Side: BUY')
        print('Tauros bid order price: ', tauros_price)
        print('Bitso bid order price: ', bitso_price)
        print("Buy order successfully placed: ")
        order_data = order_placed['data']
        real_spread = (bitso_price - Decimal(order_data['price'])) / bitso_price
        real_spread = abs(round(real_spread * 100, 2))
        print(
            f"DT: {order_data['created_at']} | Price: {order_data['price']} | Spread: {real_spread}% | Amount: {order_data['amount']} | Value: {order_data['value']}"
        )
        print(f"Sleeping {time_to_sleep} seconds")
        time.sleep(time_to_sleep)

        close_order = tauros.close_order(order_id)
        if not close_order['success']:
            logging.error('Order close failed')
            logging.error(close_order['msg'])


if __name__ == '__main__':
    print("Environment: ", "PRODUCTION" if is_production else "STAGING")

    close_all_orders()
    from multiprocessing import Process
    processes = []

    for i in range(0, 50):
        if settings.USE_FIREBASE:
            print('Using firebase...')
            response = requests.get(f'https://{settings.FIREBASE_PROJECT_ID}.firebaseio.com/{i}.json')
            robot = response.json()
        else:
            print('Using robots.json file...')
            with open('./robots.json') as bots_config:
                try:
                    robot = json.load(bots_config)[i]
                except IndexError:
                    break
        if robot is None:
            break
        processes.append(
            Process(
                target=buy_bot if robot['side'] == 'buy' else sell_bot,
                args=(i,settings.USE_FIREBASE)
            )
        )
    try:
        for process in processes:
            time.sleep(0.1)
            process.start()
        for process in processes:
            process.join()
    except KeyboardInterrupt:
        close_all_orders()
        for process in processes:
            process.terminate()
